File: src/ocaml/bst/bst.ml
(* Works only in OCaml 4.03.0 or above *)

(* Defining the shape of trees *)
type tree =
    Lf       (* Leaf *)
  | Br of {  (* Branch *)
      left:  tree;
      value: int;
      right: tree;
    }

(* (Recursive) function find, which returns whether given integer n exists in BST t *)
let rec find t n =
  match t with
    Lf -> false
  | Br {left=l; value=v; right=r} ->
     if n = v then true
     else if n < v then find l n
     else find r n

(* (Recursive) function insert, which, given BST t and a new element n, returns 
   a new binary search tree with n *)
let rec insert t n =
  match t with
    Lf -> Br {left=Lf; value=n; right=Lf}
  | Br {left=l; value=v; right=r} ->
     if n = v then t
     else if n < v then Br {left=insert l n; value=v; right=r}
     else Br {left=l; value=v; right=insert r n}

(* Function min, which, given BST t, returns the minimum value stored in t.
   If t is empty, it returns -255. *)
let rec min t =
  match t with
    Lf -> -255
  | Br {left=Lf; value=v; right=_} -> v
  | Br {left=l; value=_; right=_} -> min l
      
(* (Recursive) function delete, which, given BST t and an element n to
   be deleted, returns a new binary search tree without n.  If n is not
   stored in t, it returns t as it is. *)

let rec delete t n =
  match t with
    Lf -> t
  | Br {left=l; value=v; right=r} ->
     if n = v then 
       match l, r with
	 Lf, Lf -> Lf
       | Br _, Lf -> l
       | Lf, Br _ -> r
       | Br _, Br _ ->
	  let m = min r in
	  Br {left=l; value=m; right=delete r m}
     else if n < v then Br {left=delete l n; value=v; right=r}
     else Br {left=l; value=v; right=delete r n}

(* Constructing a sample tree *)
let t1 = Br {left = Lf; value = 10; right = Lf}
let t2 = Br {left = Lf; value = 25; right = Lf}
let t3 = Br {left = t1; value = 15; right = t2}
let t4 = Br {left = Lf; value = 60; right = Lf}
let t5 = Br {left = Lf; value = 48; right = t4}
let t6 = Br {left = t3; value = 30; right = t5}

(* Testing find *)
let test1 = find t6 30  (* should be true *)
let test2 = find t6 13  (* should be false *)

(* Testing insert *)
let t7 = insert t6 23
let t8 = insert t6 0
let test3 = find t7 23  (* should return true *)
let test4 = find t8 30  (* should return false *)
let test5 = find t8 23  (* should return false *)

(* Testing delete *)
let t9 = delete t8 30
let test6 = find t9 30
let test7 = find t9 48
	     
module Aux =
struct
  let rec preorder t =
    match t with
      Lf -> []
    | Br {left=l; value=v; right=r} ->
       preorder l @ v :: preorder r
end

File: src/C/bst/bst.c
/* Immutable binary search tree in C
 *  by Atsushi Igarashi
 *  on August 31, 2016
 */

#include <stdio.h>   // for printf
#include <stdlib.h>  // for malloc
#include <stdbool.h> // for type bool, true, and false

struct tree {
  enum nkind { LEAF, BRANCH } tag;
  union {
    struct leaf {} lf;
    struct branch {
      struct tree *left;
      int value;
      struct tree *right;
    } br;
  } dat;  // standing for DATum
};

/* Because the union member lf is never used and there is no point in
   using union in this case, we could simplify the type definition as
   follows:

struct tree {
  enum nkind { LEAF, BRANCH } tag;
  struct branch {
    struct tree *left;
    int value;
    struct tree *right;
  } br;
};

   but, for uniformity, a more complicated version is used
*/

struct tree *newbranch(struct tree *_left, int _value, struct tree *_right) {
  // Allocate a new object in the heap
  struct tree *n = (struct tree *)malloc(sizeof(struct tree));
  // And then initialize the members
  n->tag = BRANCH;            // could be written (*n).tag = BRANCH
  n->dat.br.left = _left;
  n->dat.br.value = _value;
  n->dat.br.right = _right;
  return n;
}

struct tree *newleaf() {
  struct tree *n = (struct tree *)malloc(sizeof(struct tree));
  n->tag = LEAF;
  return n;
}

bool find(struct tree *t, int n) {
  if (t->tag == LEAF) {
    return false;
  } else /* t->tag == BRANCH */ {
    struct branch b = t->dat.br;
    if (n == b.value) {
      return true;
    } else if (n < b.value) {
      return find(b.left, n);
    } else /* n > b.value */ {
      return find(b.right, n);
    }
  }
}

struct tree *insert(struct tree *t, int n) {
  if (t->tag == LEAF) {
    return newbranch(newleaf(), n, newleaf());
  } else /* t->tag == BRANCH */ {
    struct branch b = t->dat.br;
    if (n == b.value) {
      return t;
    } else if (n < b.value) {
      return newbranch(insert(b.left, n), b.value, b.right);
    } else /* n > b.value */ {
      return newbranch(b.right, b.value, insert(b.right, n));
    }
  }
}

int min(struct tree *t) {
  if (t->tag == LEAF) {
    return -255;
  } else /* t->tag == BRANCH */ {
    struct branch b = t->dat.br;
    if (b.left->tag == LEAF) {
      return b.value;
    } else {
      return min(b.left);
    }
  }
}

struct tree *delete(struct tree *t, int n) {
  if (t->tag == LEAF) {
    return t;
  } else /* t->tag == BRANCH */ {
    struct branch b = t->dat.br;
    if (n == b.value) {
      if (b.left->tag == LEAF) {
        if (b.right->tag == LEAF) {
          return newleaf();
        } else /* b.right->tag == BRANCH*/ {
          return b.right;
        }
      } else /* b.left->tag == BRANCH*/ {
        if (b.right->tag == LEAF) {
          return b.left;
        } else /* b.right->tag == BRANCH*/ {
          int m = min(b.right);
          struct tree *newRight = delete(b.right, m);
          return newbranch(b.left, m, newRight);
        }
      }
    } else if (n < b.value) {
      struct tree *newLeft = delete(b.left, n);
      return newbranch(newLeft, b.value, b.right);
    } else /* n > b.value */ {
      struct tree *newRight = delete(b.right, n);
      return newbranch(b.left, b.value, newRight);
    }
  }
}

int main() {
  struct tree *t1 = newbranch(newleaf(), 10, newleaf());
  struct tree *t2 = newbranch(newleaf(), 25, newleaf());
  struct tree *t3 = newbranch(t1, 15, t2);
  struct tree *t4 = newbranch(newleaf(), 60, newleaf());
  struct tree *t5 = newbranch(newleaf(), 48, t4);
  struct tree *t6 = newbranch(t3, 30, t5);
  bool test1 = find(t6, 30);  // should be true
  bool test2 = find(t6, 13);  // should be false
  struct tree *t7 = insert(t6, 23);
  struct tree *t8 = insert(t6, 0);
  bool test3 = find(t7,23);  // should be true
  bool test4 = find(t8, 30);  // should be true
  bool test5 = find(t8, 23);  // should be false
  struct tree *t9 = delete(t8, 30);
  bool test6 = find(t9, 30);  // should be false
  bool test7 = find(t9,48);  // should be true
  printf("test 1: %d\n", test1);
  printf("test 2: %d\n", test2);
  printf("test 3: %d\n", test3);
  printf("test 4: %d\n", test4);
  printf("test 5: %d\n", test5);
  printf("test 6: %d\n", test6);
  printf("test 7: %d\n", test7);
  return 0;
}
